File: frontend/src/api/resourceService.js
import axios from 'axios';

const API_BASE = import.meta.env.VITE_API_V1_BASE_URL || 'http://10.50.20.47:8081/api/v1';

const resourceService = {

  // Get all resources with optional filters
  getAllResources: async (filters = {}) => {
    const params = new URLSearchParams();
    if (filters.type) params.append('type', filters.type);
    if (filters.location) params.append('location', filters.location);
    if (filters.status) params.append('status', filters.status);
    if (filters.minCapacity) params.append('minCapacity', filters.minCapacity);
    const response = await axios.get(`${API_BASE}/resources?${params}`);
    return response.data;
  },

  // Get single resource by ID
  getResourceById: async (id) => {
    const response = await axios.get(`${API_BASE}/resources/${id}`);
    return response.data;
  },

  // Create new resource (Admin)
  createResource: async (resourceData) => {
    const response = await axios.post(`${API_BASE}/resources`, resourceData);
    return response.data;
  },

  // Update resource (Admin)
  updateResource: async (id, resourceData) => {
    const response = await axios.put(`${API_BASE}/resources/${id}`, resourceData);
    return response.data;
  },

  // Update status only (Admin)
  updateStatus: async (id, status) => {
    const response = await axios.patch(
      `${API_BASE}/resources/${id}/status?status=${status}`
    );
    return response.data;
  },

  // Delete resource (Admin)
  deleteResource: async (id) => {
    await axios.delete(`${API_BASE}/resources/${id}`);
  },

};

export default resourceService;

File: backend/src/main/java/com/example/backend/service/ResourceService.java
package com.example.backend.service;

import com.example.backend.dto.ResourceDTO;
import com.example.backend.model.Resource;
import com.example.backend.model.ResourceStatus;
import com.example.backend.model.ResourceType;
import com.example.backend.repository.ResourceRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class ResourceService {

    private final ResourceRepository resourceRepository;

    // Get all resources
    public List<Resource> getAllResources() {
        return resourceRepository.findAll();
    }

    // Get resource by ID
    public Optional<Resource> getResourceById(Long id) {
        return resourceRepository.findById(id);
    }

    // Search by type
    public List<Resource> getResourcesByType(ResourceType type) {
        return resourceRepository.findByType(type);
    }

    // Search by location
    public List<Resource> getResourcesByLocation(String location) {
        return resourceRepository.findByLocationContainingIgnoreCase(location);
    }

    // Search by status
    public List<Resource> getResourcesByStatus(ResourceStatus status) {
        return resourceRepository.findByStatus(status);
    }

    // Search by minimum capacity
    public List<Resource> getResourcesByMinCapacity(Integer capacity) {
        return resourceRepository.findByCapacityGreaterThanEqual(capacity);
    }

    // Create new resource (Admin)
    public Resource createResource(ResourceDTO dto) {
        if (resourceRepository.findByCode(dto.getCode()).isPresent()) {
            throw new org.springframework.web.server.ResponseStatusException(
                org.springframework.http.HttpStatus.CONFLICT, 
                "Resource with code " + dto.getCode() + " already exists."
            );
        }
        Resource resource = new Resource();
        resource.setName(dto.getName());
        resource.setCode(dto.getCode());
        resource.setType(dto.getType());
        resource.setCapacity(dto.getCapacity());
        resource.setLocation(dto.getLocation());
        resource.setAvailabilityWindow(dto.getAvailabilityWindow());
        resource.setStatus(dto.getStatus() != null ? dto.getStatus() : ResourceStatus.ACTIVE);
        resource.setDescription(dto.getDescription());
        return resourceRepository.save(resource);
    }

    // Update resource (Admin)
    public Optional<Resource> updateResource(Long id, ResourceDTO dto) {
        return resourceRepository.findById(id).map(resource -> {
            resource.setName(dto.getName());
            resource.setCode(dto.getCode());
            resource.setType(dto.getType());
            resource.setCapacity(dto.getCapacity());
            resource.setLocation(dto.getLocation());
            resource.setAvailabilityWindow(dto.getAvailabilityWindow());
            resource.setStatus(dto.getStatus());
            resource.setDescription(dto.getDescription());
            return resourceRepository.save(resource);
        });
    }

    // Update status only (Admin)
    public Optional<Resource> updateStatus(Long id, ResourceStatus status) {
        return resourceRepository.findById(id).map(resource -> {
            resource.setStatus(status);
            return resourceRepository.save(resource);
        });
    }

    // Delete resource (Admin)
    public boolean deleteResource(Long id) {
        if (resourceRepository.existsById(id)) {
            resourceRepository.deleteById(id);
            return true;
        }
        return false;
    }
}